src/dotnet/Notification.UI.Blazor/NotificationUI.cs

```diff
@@ -132,29 +132,70 @@ public void SetPermissionState(string permissionState)
         SetIsGranted(state);
     }
 
+    public async Task DeregisterDevice(CancellationToken cancellationToken = default)
+    {
+        var deviceId = await DeviceTokenRetriever.GetDeviceToken(cancellationToken).ConfigureAwait(false);
+        if (deviceId == null)
+            return;
+
+        var deleteTokenTask = DeviceTokenRetriever.DeleteDeviceToken(cancellationToken);
+        var command = new Notifications_DeregisterDevice(Session, deviceId);
+        var deregisterDeviceTask =  Hub.Commander().Call(command, cancellationToken);
+        await Task.WhenAll(deleteTokenTask, deregisterDeviceTask).ConfigureAwait(false);
+    }
+
+    public async Task EnsureDeviceRegistered(CancellationToken cancellationToken = default)
+    {
+        var deviceId = await DeviceTokenRetriever.GetDeviceToken(cancellationToken).ConfigureAwait(false);
+        var existingTask = _registerDeviceTask;
+        if (existingTask != null) {
+            var alreadyRegisteredDeviceId = await existingTask.ConfigureAwait(false);
+            if (alreadyRegisteredDeviceId == deviceId)
+                return;
+        }
+        _registerDeviceTask = null;
```

Comment:
you need to synchronize write access to _registerDeviceTask field

Reply:
OK. it will be simpler